src/terms.ml

```diff
@@ -203,6 +297,50 @@ type term =
     can hence be substituted in [r.rhs] with [Bindlib.msubst r.rhs env] to get
     the result of the application of the rule. *)
 
+(** {3 Decision trees for rewriting} *)
+
+(** Trees are used to efficiently choose a rewriting rule given a list of
+    terms (beginning with a symbol) to be rewritten.  The left hand has side
+    of the rule is spread across the {!constructor:Node}s of the tree.  Hence,
+    progressing down the tree is equivalent to reducing the set of possible
+    rules.  When a {!constructor:Leaf} is reached, the target is rewritten to
+    the content of the leaf. *)
+ and tree =
+  | Leaf of int IntMap.t * (term_env, term) Bindlib.mbinder
+  (** Holds the right hand side of a rule.  In a {!constructor:Leaf}[(e, a)],
+      - [a] is the right hand side of the rule.
+      - [e] maps a position in the stack containing terms which stand as
+            pattern variables in some rules to the slot allocated in the
+            {!type:term_env array}. *)
+  | Node of node_data
+  (** Nodes allow to perform switches, a switch being the matching of a
+      pattern.  Briefly, a {!constructor:Node} contains one subtree per
+      possible switch, plus possibly a default case and an abstraction
+      case. *)
+  | Fetch of (bool * tree)
```

Comment:
"Fetch" seems to be a new constructor. Could you give a more detailed explanation? What is the boolean argument?